### ISSUE TITLE
KIS tasks in 0.15.1 RC2 sometimes duplicate rows with the same dimension values

### ISSUE THREAD

**Issue by pdeva** [apache/incubator-druid#8276]
### Affected Version

0.15.1 RC2

### Description

CSV of Results:
[query-943282d6-3b66-4881-8080-355b09600870.zip](https://github.com/apache/incubator-druid/files/3484553/query-943282d6-3b66-4881-8080-355b09600870.zip)

Note that there are 2 rows for timestamp `2019-08-09T02:38:00.000Z`, even though the dimensions have the exact same value. In fact, for some reason, the rows are fully duplicated for that time period.

This reflects in charts as values are doubled and it happens pretty often. We noticed it when we saw random spikes in charts when using the aggregated data source. These spikes happen whenever druid duplicates the rows

![image](https://user-images.githubusercontent.com/3190176/62751456-95664600-ba18-11e9-9588-dab0e23fcd6a.png)

SQL

``` SQL
SELECT *
 FROM "kube-metrics-minute" 
 WHERE accountid='52c7e28eebb26a86fb1505b5' AND  
 "__time" >= TIMESTAMP '2019-08-09 02:37:00' AND "__time" < TIMESTAMP '2019-08-09 02:40:00'  
 AND cluster='ds' AND etype='container' AND namespace='default' AND name='app-alertproc' AND pod='alertproc-deployment-65f59b6b8-gc57m' 
```

Generated query (shown in router web console)

```js
{
  "queryType": "scan",
  "dataSource": {
    "type": "table",
    "name": "kube-metrics-minute"
  },
  "intervals": {
    "type": "intervals",
    "intervals": [
      "2019-08-09T02:37:00.000Z/2019-08-09T02:40:00.000Z"
    ]
  },
  "virtualColumns": [],
  "resultFormat": "compactedList",
  "batchSize": 20480,
  "limit": 9223372036854776000,
  "order": "none",
  "filter": {
    "type": "and",
    "fields": [
      {
        "type": "selector",
        "dimension": "accountid",
        "value": "52c7e28eebb26a86fb1505b5",
        "extractionFn": null
      },
      {
        "type": "selector",
        "dimension": "cluster",
        "value": "ds",
        "extractionFn": null
      },
      {
        "type": "selector",
        "dimension": "etype",
        "value": "container",
        "extractionFn": null
      },
      {
        "type": "selector",
        "dimension": "namespace",
        "value": "default",
        "extractionFn": null
      },
      {
        "type": "selector",
        "dimension": "name",
        "value": "app-alertproc",
        "extractionFn": null
      },
      {
        "type": "selector",
        "dimension": "pod",
        "value": "alertproc-deployment-65f59b6b8-gc57m",
        "extractionFn": null
      }
    ]
  },
  "columns": [
    "__time",
    "accountid",
    "alloc_cpu",
    "alloc_mem",
    "alloc_pods",
    "cap_cpu",
    "cap_mem",
    "cap_pods",
    "cluster",
    "count",
    "cpu",
    "etype",
    "image",
    "labels",
    "lim_cpu",
    "lim_mem",
    "mem_limit",
    "mem_used",
    "name",
    "namespace",
    "node",
    "node_pool",
    "owner",
    "owner_kind",
    "pod",
    "req_cpu",
    "req_mem",
    "rxBytes",
    "txBytes"
  ],
  "legacy": false,
  "context": {
    "sqlQueryId": "f39f0b78-bb97-4c81-b9dd-7e0e2abf0b46"
  },
  "descending": false,
  "granularity": {
    "type": "all"
  }
}
```

Supervisor task spec:

```js
{
  "type": "kafka",
  "dataSchema": {
    "dataSource": "kube-metrics-minute",
    "parser": {
      "type": "string",
      "parseSpec": {
        "format": "json",
        "timestampSpec": {
          "column": "timestamp",
          "format": "auto"
        },
        "dimensionsSpec": {
          "dimensions": [
            "accountid",
            "cluster",
            "name",
            "namespace",
            "etype",
            "labels",
            "owner",
            "owner_kind",
            "node",
            "image",
            "pod",
            "node_pool"
          ]
        }
      }
    },
    "metricsSpec": [
      {
        "name": "count",
        "type": "count"
      },
      {
        "name": "req_cpu",
        "fieldName": "req_cpu",
        "type": "floatMax"
      },
      {
        "name": "req_mem",
        "fieldName": "req_mem",
        "type": "longMax"
      },
      {
        "name": "lim_cpu",
        "fieldName": "lim_cpu",
        "type": "floatMax"
      },
      {
        "name": "lim_mem",
        "fieldName": "lim_mem",
        "type": "longMax"
      },
      {
        "name": "cap_cpu",
        "fieldName": "cap_cpu",
        "type": "floatMax"
      },
      {
        "name": "cap_mem",
        "fieldName": "cap_mem",
        "type": "longMax"
      },
      {
        "name": "cap_pods",
        "fieldName": "cap_pods",
        "type": "floatMax"
      },
      {
        "name": "alloc_cpu",
        "fieldName": "alloc_cpu",
        "type": "floatMax"
      },
      {
        "name": "alloc_mem",
        "fieldName": "alloc_mem",
        "type": "longMax"
      },
      {
        "name": "alloc_pods",
        "fieldName": "alloc_pods",
        "type": "floatMax"
      },
      {
        "name": "cpu",
        "fieldName": "cpu",
        "type": "doubleMax"
      },
      {
        "name": "rxBytes",
        "fieldName": "rxBytes",
        "type": "longMax"
      },
      {
        "name": "txBytes",
        "fieldName": "txBytes",
        "type": "longMax"
      },
      {
        "name": "mem_used",
        "fieldName": "mem_used",
        "type": "longMax"
      },
      {
        "name": "mem_limit",
        "fieldName": "mem_limit",
        "type": "longMax"
      }
    ],
    "granularitySpec": {
      "type": "uniform",
      "segmentGranularity": {"type": "period", "period": "PT4H"},
      "queryGranularity": "MINUTE"
    }
  },
  "tuningConfig": {
    "type": "kafka",
    "maxRowsPerSegment": 5000000
  },
  "ioConfig": {
    "topic": "kube-metrics",
    "consumerProperties": {
      "bootstrap.servers": "xx"
    },
    "taskCount": 1,
    "replicas": 2,
    "taskDuration": "PT4H"
  }
}

```

**Comment by vogievetsky** [comment]
Is it possible that those rows are duplicated in Kafka?

**Comment by pdeva** [comment]
nope. in fact, we have another datasource reading from the same topic with 10s granularity, and it shows no duplication at all. (10s is also the granularity at which data is generated).

Results from different data source with higher resolution (base resolution at which data is generated)
[query-3014d0a9-3a96-4b2f-9180-b90a8c73a1f1.zip](https://github.com/apache/incubator-druid/files/3484776/query-3014d0a9-3a96-4b2f-9180-b90a8c73a1f1.zip)
Note that data here is perfect for the same time range and same kafka topic

```SQL
SELECT *
 FROM "kube-metrics-second" 
 WHERE accountid='52c7e28eebb26a86fb1505b5' AND  
 "__time" >= TIMESTAMP '2019-08-09 02:37:00' AND "__time" < TIMESTAMP '2019-08-09 02:40:00'  
 AND cluster='ds' AND etype='container' AND namespace='default' AND name='app-alertproc' AND pod='alertproc-deployment-65f59b6b8-gc57m'
```
KIS spec

```js
{
  "type": "kafka",
  "dataSchema": {
    "dataSource": "kube-metrics-second",
    "parser": {
      "type": "string",
      "parseSpec": {
        "format": "json",
        "timestampSpec": {
          "column": "timestamp",
          "format": "auto"
        },
        "dimensionsSpec": {
          "dimensions": [
            "accountid",
            "cluster",
            "name",
            "namespace",
            "etype",
            "labels",
            "owner",
            "owner_kind",
            "node",
            "image",
            "pod",
            "node_pool"
          ]
        }
      }
    },
    "metricsSpec": [
      {
        "name": "count",
        "type": "count"
      },
      {
        "name": "req_cpu",
        "fieldName": "req_cpu",
        "type": "floatMax"
      },
      {
        "name": "req_mem",
        "fieldName": "req_mem",
        "type": "longMax"
      },
      {
        "name": "lim_cpu",
        "fieldName": "lim_cpu",
        "type": "floatMax"
      },
      {
        "name": "lim_mem",
        "fieldName": "lim_mem",
        "type": "longMax"
      },
      {
        "name": "cap_cpu",
        "fieldName": "cap_cpu",
        "type": "floatMax"
      },
      {
        "name": "cap_mem",
        "fieldName": "cap_mem",
        "type": "longMax"
      },
      {
        "name": "cap_pods",
        "fieldName": "cap_pods",
        "type": "floatMax"
      },
      {
        "name": "alloc_cpu",
        "fieldName": "alloc_cpu",
        "type": "floatMax"
      },
      {
        "name": "alloc_mem",
        "fieldName": "alloc_mem",
        "type": "longMax"
      },
      {
        "name": "alloc_pods",
        "fieldName": "alloc_pods",
        "type": "floatMax"
      },
      {
        "name": "cpu",
        "fieldName": "cpu",
        "type": "doubleMax"
      },
      {
        "name": "rxBytes",
        "fieldName": "rxBytes",
        "type": "longMax"
      },
      {
        "name": "txBytes",
        "fieldName": "txBytes",
        "type": "longMax"
      },
      {
        "name": "mem_used",
        "fieldName": "mem_used",
        "type": "longMax"
      },
      {
        "name": "mem_limit",
        "fieldName": "mem_limit",
        "type": "longMax"
      }
    ],
    "granularitySpec": {
      "type": "uniform",
      "segmentGranularity": "HOUR",
      "queryGranularity": "SECOND"
    }
  },
  "tuningConfig": {
    "type": "kafka",
    "maxRowsPerSegment": 5000000
  },
  "ioConfig": {
    "topic": "kube-metrics",
    "consumerProperties": {
      "bootstrap.servers": "xxx"
    },
    "taskCount": 1,
    "replicas": 2,
    "taskDuration": "PT1H"
  }
}

```

**Comment by pdeva** [comment]
it gets worse.

[query-a65b3b59-3a39-4168-9261-912020fd591c.zip](https://github.com/apache/incubator-druid/files/3487289/query-a65b3b59-3a39-4168-9261-912020fd591c.zip)

here the aggregation is PT30M, and you can see `2019-08-09T15:00:00.000Z` row is repeated four times!!

```sql
SELECT *
 FROM "kube-metrics-30m" 
 WHERE accountid='52c7e28eebb26a86fb1505b5' AND  
 "__time" >= TIMESTAMP '2019-08-09 14:00:00' AND "__time" < TIMESTAMP '2019-08-09 16:30:00'  
 AND cluster='ds' AND etype='container' AND namespace='default' AND name='app-alertproc' AND pod='alertproc-deployment-65f59b6b8-gc57m'
 ```

KIS spec

```
{
  "type": "kafka",
  "dataSchema": {
    "dataSource": "kube-metrics-30m",
    "parser": {
      "type": "string",
      "parseSpec": {
        "format": "json",
        "timestampSpec": {
          "column": "timestamp",
          "format": "auto"
        },
        "dimensionsSpec": {
          "dimensions": [
            "accountid",
            "cluster",
            "name",
            "namespace",
            "etype",
            "labels",
            "owner",
            "owner_kind",
            "node",
            "image",
            "pod",
            "node_pool"
          ]
        }
      }
    },
    "metricsSpec": [
      {
        "name": "count",
        "type": "count"
      },
      {
        "name": "req_cpu",
        "fieldName": "req_cpu",
        "type": "floatMax"
      },
      {
        "name": "req_mem",
        "fieldName": "req_mem",
        "type": "longMax"
      },
      {
        "name": "lim_cpu",
        "fieldName": "lim_cpu",
        "type": "floatMax"
      },
      {
        "name": "lim_mem",
        "fieldName": "lim_mem",
        "type": "longMax"
      },
      {
        "name": "cap_cpu",
        "fieldName": "cap_cpu",
        "type": "floatMax"
      },
      {
        "name": "cap_mem",
        "fieldName": "cap_mem",
        "type": "longMax"
      },
      {
        "name": "cap_pods",
        "fieldName": "cap_pods",
        "type": "floatMax"
      },
      {
        "name": "alloc_cpu",
        "fieldName": "alloc_cpu",
        "type": "floatMax"
      },
      {
        "name": "alloc_mem",
        "fieldName": "alloc_mem",
        "type": "longMax"
      },
      {
        "name": "alloc_pods",
        "fieldName": "alloc_pods",
        "type": "floatMax"
      },
      {
        "name": "cpu",
        "fieldName": "cpu",
        "type": "doubleMax"
      },
      {
        "name": "rxBytes",
        "fieldName": "rxBytes",
        "type": "longMax"
      },
      {
        "name": "txBytes",
        "fieldName": "txBytes",
        "type": "longMax"
      },
      {
        "name": "mem_used",
        "fieldName": "mem_used",
        "type": "longMax"
      },
      {
        "name": "mem_limit",
        "fieldName": "mem_limit",
        "type": "longMax"
      }
    ],
    "granularitySpec": {
      "type": "uniform",
      "segmentGranularity": {"type": "period", "period": "PT12H"},
      "queryGranularity": {"type": "period", "period": "PT30M"}
    }
  },
  "tuningConfig": {
    "type": "kafka",
    "maxRowsPerSegment": 5000000
  },
  "ioConfig": {
    "topic": "kube-metrics",
    "consumerProperties": {
      "bootstrap.servers": "xx"
    },
    "taskCount": 1,
    "replicas": 2,
    "taskDuration": "PT12H"
  }
}
```

**Comment by pdeva** [comment]
adding more data to help getting to root cause.

24 hour of data when queried from `minute` granularity datasource (`kube-metrics-minute` spec above)

![image](https://user-images.githubusercontent.com/3190176/62800349-0d278580-ba98-11e9-873c-0b01715eaef3.png)

Same data, when queried from `second` granularity datasource (`kube-metrics-second` above)
![image](https://user-images.githubusercontent.com/3190176/62800412-321bf880-ba98-11e9-9c76-e866cfba2170.png)

**Comment by jon-wei** [comment]
With Kafka ingestion, it's possible that rollup is not 100%, due to the incremental publishing, e.g.:
- Kafka task gets a row like `{"timestamp":"2019-08-09 14:00:00", "dim1":"hello", "metric":1}`
- Kafka task hits a memory/row/time limit and does an incremental publish
- If after the incremental publish, another row `{"timestamp":"2019-08-09 14:00:00", "dim1":"hello", "metric":2}` arrives, it won't be rolled up into the first already published segment, and a non-aggregating query would return both of those rows

If you run an exact aggregating query like:

```
SELECT accountid, SUM(cpu) 
FROM "kube-metrics-minute" 
WHERE accountid='52c7e28eebb26a86fb1505b5' AND  
 "__time" >= TIMESTAMP '2019-08-09 14:00:00' AND "__time" < TIMESTAMP '2019-08-09 16:30:00' AND cluster='ds' AND etype='container' AND namespace='default' AND name='app-alertproc' AND pod='alertproc-deployment-65f59b6b8-gc57m' 
GROUP BY accountid
```

does the sum on CPU return the expected value?

You could compare the results of that query across the minute/second/30m datasources and see if they match, to rule out imperfect rollup from incremental publishes.

**Comment by pdeva** [comment]
they all produce different value.
here is the video showing results on each datasource:
https://www.screencast.com/t/gSsZsnHz0

as for the points you mentioned, they are not possible due to the very small amount of data that is being fed into this kafka topic and the frequency with which the issue is being seen.
I described in the druid-dev thread `ingestion time aggregation in KIS broken in 0.15.1`

**Comment by jon-wei** [comment]
Ah, sorry, I just realized that "cpu" is a max aggregator here, can you try a sum on count instead?

```
SELECT accountid, SUM(count) 
FROM "kube-metrics-minute" 
WHERE accountid='52c7e28eebb26a86fb1505b5' AND  
 "__time" >= TIMESTAMP '2019-08-09 14:00:00' AND "__time" < TIMESTAMP '2019-08-09 16:30:00' AND cluster='ds' AND etype='container' AND namespace='default' AND name='app-alertproc' AND pod='alertproc-deployment-65f59b6b8-gc57m' 
GROUP BY accountid
```

**Comment by jon-wei** [comment]
If I compare the 1m vs 1s data provided, they look consistent:

1s for 2019-08-09T02:37:00.000Z:

![Screen Shot 2019-08-09 at 1 40 36 PM](https://user-images.githubusercontent.com/8729063/62807904-82e91c80-baab-11e9-849b-db0eb780f2da.png)

1s for 2019-08-09T02:38:00.000Z:

![Screen Shot 2019-08-09 at 1 43 11 PM](https://user-images.githubusercontent.com/8729063/62807956-a613cc00-baab-11e9-9761-2f0048ff1b79.png)

1m data:




![Screen Shot 2019-08-09 at 1 41 00 PM](https://user-images.githubusercontent.com/8729063/62807928-91373880-baab-11e9-93d3-60cdbbe69abe.png)

The 1m data looks consistent with the 1s data, with 6 total rows from the input, and the max agg on CPU is consistent as well

**Comment by pdeva** [comment]
you are right, the sum of counts is indeed the same across datasources.
video:
https://www.screencast.com/t/lRPk0aif6qZ

However, looking at the segments for the `30m` datasource, they are clearly too small to hit any kind of memory/row limit. So why is there any duplication of rows at all?

![image](https://user-images.githubusercontent.com/3190176/62808527-2e46a100-baad-11e9-97d2-0689c37ad078.png)

**Comment by jon-wei** [comment]
Hm, are you able to provide a task log for the 30m task?

**Comment by pdeva** [comment]
what is a 'task log' and where do i get it from?

**Comment by jon-wei** [comment]
in the tasks view, if you click the magnifying glass next to a task, it will have a `Log` section on the left in the window that pops up, grab the `View full log`

**Comment by pdeva** [comment]
nothing in the logs. just `Finished peon task`

**Comment by jon-wei** [comment]
I think that means log4j logging isn't configured (the `Finished peon task` is printed using `System.out.println` instead of through a logger), check https://druid.apache.org/docs/latest/configuration/logging.html

**Comment by pdeva** [comment]
this is how logging is configured for all nodes.
apart from the peons, all the other nodes seem to be outputting fine..

```xml
<Configuration status="WARN">
    <Appenders>

        <RollingFile name="RollingFile" fileName="log/druid.log" filePattern="log/druid-%d{MM-dd-yyyy}-%i.log">

          <PatternLayout>
                <Pattern>%d %p %c{1.} [%t] %m%n</Pattern>
          </PatternLayout>
          <Policies>
                <OnStartupTriggeringPolicy />
                <TimeBasedTriggeringPolicy />
                <SizeBasedTriggeringPolicy size="100 MB"/>
          </Policies>

          <DefaultRolloverStrategy>
            <Delete basePath="log/">
              <IfFileName glob="*/druid-*.log">
                <IfLastModified age="7d">
                  <IfAny>
                    <IfAccumulatedFileSize exceeds="4 GB" />
                    <IfAccumulatedFileCount exceeds="10" />
                  </IfAny>
                </IfLastModified>
              </IfFileName>
              
            </Delete>
          </DefaultRolloverStrategy>
          

        </RollingFile>


    </Appenders>

    <Loggers>
        
        <AsyncLogger name="io.druid.curator.inventory.CuratorInventoryManager" level="info" additivity="false">
          <AppenderRef ref="RollingFile"/>
        </AsyncLogger>

        <AsyncLogger name="io.druid.client.BatchServerInventoryView" level="info" additivity="false">
          <AppenderRef ref="RollingFile"/>
        </AsyncLogger>

        <!-- Make extra sure nobody adds logs in a bad way that can hurt performance -->
        <AsyncLogger name="io.druid.client.ServerInventoryView" level="info" additivity="false">
          <AppenderRef ref="RollingFile"/>
        </AsyncLogger>
        <AsyncLogger name ="io.druid.java.util.http.client.pool.ChannelResourceFactory" level="info" additivity="false">
          <AppenderRef ref="RollingFile"/>
        </AsyncLogger>

        <Root level="info">
            <AppenderRef ref="RollingFile"/>
        </Root>
    </Loggers>
</Configuration>

```

**Comment by jon-wei** [comment]
In the middleManager runtime.properties, does the `druid.indexer.runner.javaOpts` property have the following? 

`-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager`

It will be needed for task logging to work.

Additionally, I think you'll need a separate log4j xml for the tasks that uses a Console appender (the separate task logs rely on the tasks logging to console), you can specify that with `-Dlog4j.configurationFile=<conf-file-location>`, this should also go in `druid.indexer.runner.javaOpts` in the MM runtime properties

**Comment by pdeva** [comment]
I do have the `-Djava.util.logging.manager=org.apache.logging.log4j.jul.LogManager` property in `druid.indexer.runner.javaOpts`.

I will add the `-Dlog4j.configurationFile=<conf-file-location>` with the following config, and let you know what happens.  If you need me to add additional config to the below let me know

```xml

<?xml version="1.0" encoding="UTF-8" ?>

<Configuration status="WARN">
    <Appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="%d{ISO8601} %p [%t] %c - %m%n"/>
        </Console>
    </Appenders>

    <Loggers>

        <AsyncLogger name="io.druid.curator.inventory.CuratorInventoryManager" level="info" additivity="false">
            <AppenderRef ref="Console"/>
        </AsyncLogger>

        <AsyncLogger name="io.druid.client.BatchServerInventoryView" level="info" additivity="false">
            <AppenderRef ref="Console"/>
        </AsyncLogger>

        <!-- Make extra sure nobody adds logs in a bad way that can hurt performance -->
        <AsyncLogger name="io.druid.client.ServerInventoryView" level="info" additivity="false">
            <AppenderRef ref="Console"/>
        </AsyncLogger>
        <AsyncLogger name ="io.druid.java.util.http.client.pool.ChannelResourceFactory" level="info" additivity="false">
            <AppenderRef ref="Console"/>
        </AsyncLogger>

        <Root level="info">
            <AppenderRef ref="Console"/>
        </Root>
    </Loggers>
</Configuration>

```

**Comment by pdeva** [comment]
ok, i enabled the task logs, let everything run for a full day and collected all the data in this zip.
[druidbug.zip](https://github.com/apache/incubator-druid/files/3490327/druidbug.zip)

the zip contains everything, the queries, the results, etc.
this time again, the `30m` segment has multiple repeated rows.

**Comment by jon-wei** [comment]
That task looks pretty normal, the spec has `"intermediatePersistPeriod" : "PT10M"`, and the log shows the task making several time-based incremental persists. 

If you're concerned about the rollup for segments generated by Kafka indexing tasks, you can run a compaction task on those intervals to fully aggregate the rows.

**Comment by pdeva** [comment]
oh i see now, the default value for intermediate persist is 10M.

So i can understand what is happening, may i ask what `intermediate persist` is? I thought it meant persisting segments.
But i don't see segments being created every 10 minutes here.

**Comment by jon-wei** [comment]
In the task logs, when you see messages like the following:

```
2019-08-11T03:54:10,121 INFO [task-runner-0-priority-0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Submitting persist runnable for dataSource[kube-metrics-30m]
2019-08-11T03:54:10,121 INFO [kube-metrics-30m-incremental-persist] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Segment[kube-metrics-30m_2019-08-11T00:00:00.000Z_2019-08-11T06:00:00.000Z_2019-08-11T00:00:00.177Z_1], persisting Hydrant[FireHydrant{, queryable=kube-metrics-30m_2019-08-11T00:00:00.000Z_2019-08-11T06:00:00.000Z_2019-08-11T00:00:00.177Z_1, count=6}]
2019-08-11T03:54:10,122 INFO [kube-metrics-30m-incremental-persist] org.apache.druid.segment.IndexMergerV9 - Starting persist for interval[2019-08-11T00:00:00.000Z/2019-08-11T06:00:00.000Z], rows[59]
```

that indicates the task is persisting in-memory segment data to disk (a `hydrant` is an intermediate persist, on disk it has the same format as any other druid segment).

The intermediate persists are queryable, and when the task decides to push to deep storage/publish, it will merge the intermediate persists into larger segments and push the merged segments, e.g.:

```
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Pushed merged index for segment[kube-metrics-30m_2019-08-11T12:00:00.000Z_2019-08-11T18:00:00.000Z_2019-08-11T12:00:00.122Z], descriptor is: DataSegment{size=24988, shardSpec=NumberedShardSpec{partitionNum=0, partitions=0}, metrics=[count, req_cpu, req_mem, lim_cpu, lim_mem, cap_cpu, cap_mem, cap_pods, alloc_cpu, alloc_mem, alloc_pods, cpu, rxBytes, txBytes, mem_used, mem_limit], dimensions=[accountid, cluster, name, namespace, etype, labels, owner, owner_kind, node, image, pod, node_pool], version='2019-08-11T12:00:00.122Z', loadSpec={type=>s3_zip, bucket=>dsprod, key=>kube-metrics-30m/2019-08-11T12:00:00.000Z_2019-08-11T18:00:00.000Z/2019-08-11T12:00:00.122Z/0/4f1a6e43-6eaf-49a8-8d12-95253b4cb203/index.zip, S3Schema=>s3n}, interval=2019-08-11T12:00:00.000Z/2019-08-11T18:00:00.000Z, dataSource='kube-metrics-30m', binaryVersion='9'}
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Pushing merged index for segment[kube-metrics-30m_2019-08-11T06:00:00.000Z_2019-08-11T12:00:00.000Z_2019-08-11T06:00:00.111Z].
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Adding hydrant[FireHydrant{, queryable=kube-metrics-30m_2019-08-11T06:00:00.000Z_2019-08-11T12:00:00.000Z_2019-08-11T06:00:00.111Z, count=0}]
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Adding hydrant[FireHydrant{, queryable=kube-metrics-30m_2019-08-11T06:00:00.000Z_2019-08-11T12:00:00.000Z_2019-08-11T06:00:00.111Z, count=1}]
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Adding hydrant[FireHydrant{, queryable=kube-metrics-30m_2019-08-11T06:00:00.000Z_2019-08-11T12:00:00.000Z_2019-08-11T06:00:00.111Z, count=2}]
2019-08-11T14:43:59,896 INFO [appenderator_merge_0] 
...
```

**Comment by repnop** [comment]
Hi, I'm having the same kind of problem and came across this issue except that in my case the data isn't duplicated, only the timestamp, and (being very new to druid) I'm not sure I can tell what the solution was here? I'm aggregating values by `longMax` for one-minute intervals and the duplicate data is causing a pretty big issue since I'm trying to determine a rate & graph it. I attempted to set the intermediate persist interval higher to see if that would help but it doesn't seem like that's the case. I also attempted to turn on compaction but I feel like I must have done that incorrectly since I don't think the task has actually run once, but since the data is real time I'm not even sure that helps in my case since we want the most up to date data as possible and having such large spikes that dwarf the rest of the graph is making it pretty unreadable (as `pandas` is adding the two values with the same timestamp together). The data is being consumed from a Kafka topic which has entries thrown onto it every ~30 seconds at the moment for testing. Any help would be appreciated, thank you!
